File: src/main/java/org/freemind/tree/MutableTreeNode.java
package org.freemind.tree;

public interface MutableTreeNode<T> extends TreeNode<T> {
  void insert(MutableTreeNode<T> child, int index);

  void remove(int index);

  void remove(MutableTreeNode<T> child);

  void setValue(T value);

  void removeFromParent();

  void setParent(MutableTreeNode<T> parent);
}

File: src/main/kotlin/org/freemind/tree/MutableTreeNode.kt
package org.freemind.tree

interface MutableTreeNode<T> : TreeNode<T> {
  fun insert(child: MutableTreeNode<T>?, index: Int)

  fun remove(index: Int)

  fun remove(child: MutableTreeNode<T>?)

  fun removeFromParent()

  override var parent: MutableTreeNode<T>?

  override var value: T
}
